app/routes.py

```diff
@@ -1 +1,147 @@
+    )
+
+
+def validate_model_id(cls, model_id):
+    try:
+        model_id = int(model_id)
+    except ValueError:
+        abort(
+            make_response(
+                f"{model_id} is an invalid input. Please input an integer.", 400
+            )
+        )
+
+    model = cls.query.get(model_id)
+    if model:
+        return model
+    else:
+        abort(make_response({"error": f"Item #{model_id} cannot be found"}, 404))
+
+
+tasks_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+
+
+@tasks_bp.route("", methods=["GET"])
+def retrieve_tasks():
+    tasks = Task.query.all()
+    tasks_response = []
+    sort_query = request.args.get("sort")
+
+    if sort_query == "asc":
+        tasks = Task.query.order_by(asc("title"))
+    elif sort_query == "desc":
+        tasks = Task.query.order_by(desc("title"))
+
+    for task in tasks:
+        tasks_response.append(task.format_task())
+    return jsonify(tasks_response)
+
+
+@tasks_bp.route("", methods=["POST"])
+def create_task():
+    request_body = request.get_json()
+
+    if not request_body.get("title") or not request_body.get("description"):
+        return {"details": "Invalid data"}, 400
+
+    new_task = Task(
+        title=request_body["title"],
+        description=request_body["description"],
+        completed_at=request_body.get("completed_at"),
+    )
+    db.session.add(new_task)
+    db.session.commit()
+    return {"task": new_task.format_task()}, 201
+
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+def handle_task(task_id):
+    task = validate_model_id(Task, task_id)
+    if task.goal_id:
+        return {"task": task.format_task_goal()}
+    else:
+        return {"task": task.format_task()}
+
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def update_task(task_id):
+    task = validate_model_id(Task, task_id)
+
+    request_body = request.get_json()
+
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+    task.completed_at = request_body.get("completed_at")
+
+    db.session.commit()
+
+    return {"task": task.format_task()}
+
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    task = validate_model_id(Task, task_id)
+
+    db.session.delete(task)
+    db.session.commit()
+
+    return make_response(
+        {"details": f'Task {task_id} "{task.title}" successfully deleted'}
+    )
+
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def update_task_as_complete(task_id):
+    task = validate_model_id(Task, task_id)
+    task.completed_at = datetime.datetime.now()
+    db.session.commit()
+
+    slackbot_call(f"Someone just completed the task {task.title}")
+
+    return make_response(
+        {
+            "task": {
+                "id": task.task_id,
+                "title": task.title,
+                "description": task.description,
+                "is_complete": (bool(task.completed_at)),
+            }
```

Comment:
I had a comment pointing out that `format_task` could have been used here, and while that's true, this could have been a downside to my suggestion that you combine `format_task_goal` when the `Task` returned contains a `"goal_id"` key (and thus possibly fail a test not expecting that key).

This could justify a separate `format_task_goal`... However, in most cases, we would still expect the `"goal_id"` key to be returned when doing this update. It's not so much that we have a reason to _not_ show that key as part of the task, it's more that we didn't have that key during the earlier waves, but now we do.

Either way, I still think the cleanest way I would implement is still handling in my suggested `format_task`, as well as using it as a helper method here. In that way, `format_task` is defining what format we always return a task in, and that stays the same across endpoints.

Reply:
The use of `format_task` also applies to `update_task_as_incomplete` down below.